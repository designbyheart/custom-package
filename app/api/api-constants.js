// @flow

export const TOKEN_EXPIRED_CODE = 'MS-107'
export const SERVER_ERROR_CODE = 'OCS-000'
export const SERVER_API_CALL_ERROR = 'OCS-001'
export const PENDING_CONNECTION_REQUEST_CODE = 'CRS-103'
export const PUSH_NOTIFICATION_SENT_CODE = 'push-notification-sent'

export const QR_CODE_SENDER_DID = 'sD'
export const QR_CODE_SENDER_VERIFICATION_KEY = 'sVk'
export const QR_CODE_LOGO_URL = 'lu'
export const QR_CODE_REQUEST_ID = 'rid'
export const QR_CODE_AGENT_PROOF = 'sakdp'
export const QR_CODE_SENDER_NAME = 'sn'
export const QR_CODE_TARGET_NAME = 'tn'
export const QR_CODE_SENDER_ENDPOINT = 'e'

export const MESSAGE_TYPE = {
  AUTH: 'auth-req',
  CLAIM_OFFER: 'claimOffer',
  CLAIM_REQUEST: 'claimReq',
  PROOF_REQUEST: 'proofReq',
  CLAIM: 'claim',
}

export const API_TYPE = {
  CONNECT: 'CONNECT',
  REGISTER: 'REGISTER',
  CREATE_AGENT: 'CREATE_AGENT',
  INVITE_ANSWERED: 'INVITE_ANSWERED',
  CREATE_KEY: 'CREATE_KEY',
}

export const PAYLOAD_TYPE = {
  GET_MESSAGE: 'GET_MSGS',
  UPDATE_PUSH_COM_METHOD: 'UPDATE_PUSH_COM_METHOD',
  SEND_MESSAGE: 'SEND_MSG',
}

export const STATUS_CODE = {
  ACCEPT: 'MS-104',
}

export const ERROR_ALREADY_EXIST = {
  code: 'OCS',
  message: 'duplicate connection request',
}

export const ERROR_INVITATION_RESPONSE_PARSE_CODE = 'RN-100'

export const ERROR_INVITATION_RESPONSE_PARSE =
  'Error parsing error message in invitation response'

export const ERROR_PENDING_INVITATION_RESPONSE_PARSE_CODE = 'RN-101'

export const ERROR_PENDING_INVITATION_RESPONSE_PARSE =
  'Error parsing error message Pending Invitation SMS'
