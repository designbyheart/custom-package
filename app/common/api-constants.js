// @flow
// TODO:KS Consider moving it to api folder
export const TOKEN_EXPIRED_CODE = 'CRS-121'
export const SERVER_ERROR_CODE = 'OCS-000'
export const SERVER_API_CALL_ERROR = 'OCS-001'
export const PENDING_CONNECTION_REQUEST_CODE = 'CRS-103'
export const PUSH_NOTIFICATION_SENT_CODE = 'push-notification-sent'

export const QR_CODE_SIGNATURE = 's'
export const QR_CODE_CHALLENGE = 'c'
export const QR_CODE_REMOTE_CONNECTION_LOGO = 'lg'
export const QR_CODE_TARGET_DID = 'tDID'
export const QR_CODE_REMOTE_HOSTING_DID = 'rhDID'
export const QR_CODE_REMOTE_PAIR_WISE_DID = 'rpDID'
export const QR_CODE_TARGET_NAME = 'tn'
export const QR_CODE_SENDER_NAME = 'sn'
export const QR_CODE_UID = 'uid'

export const PUSH_NOTIFICATION_TYPE = {
  AUTH: 'auth-req',
  CLAIM_OFFER: 'claim-offer',
}
