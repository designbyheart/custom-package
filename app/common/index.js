// @flow
export * from './api-constants'
export * from './button-constants'
export * from './message-constants'
export * from './route-constants'
export * from './secure-storage-constants'
export function noop() {}
