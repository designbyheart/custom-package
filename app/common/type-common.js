// @flow

export type CustomError = {
  code: string,
  message: string,
}

export const INITIAL_TEST_ACTION: 'INITIAL_TEST_ACTION' = 'INITIAL_TEST_ACTION'

export type InitialTestAction = {
  type: typeof INITIAL_TEST_ACTION,
}

export const initialTestAction = () => ({
  type: INITIAL_TEST_ACTION,
})

export type NavigationParams = {
  [string]: any,
}

export type ReactNavigation = {
  navigation: {
    navigate: (route: string, params?: any) => void,
    state?: {
      params: NavigationParams,
    },
  },
}

export type GenericObject = {
  [string]: any,
}

export type GenericStringObject = {
  [string]: string,
}

export type MessageAnnotation = {
  name: string,
  version: string,
}

export type TopicAnnotation = {
  tid: number,
  mid: number,
}
