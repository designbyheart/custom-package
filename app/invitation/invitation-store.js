// @flow
import { put, takeLatest, call, all, select } from 'redux-saga/effects'
import {
  INVITATION_RECEIVED,
  INVITATION_RESPONSE_SEND,
  INVITATION_RESPONSE_SUCCESS,
  INVITATION_RESPONSE_FAIL,
  INVITATION_REJECTED,
} from './type-invitation'
import { ResponseType } from '../components/request/type-request'
import {
  connectWithConsumerAgency,
  registerWithConsumerAgency,
  createAgentWithConsumerAgency,
  createAgentPairwiseKey,
  sendInvitationResponse as sendInvitationResponseApi,
} from '../api/api'
import {
  API_TYPE,
  ERROR_ALREADY_EXIST,
  ERROR_INVITATION_RESPONSE_PARSE_CODE,
  ERROR_INVITATION_RESPONSE_PARSE,
} from '../api/api-constants'
import {
  getAgencyUrl,
  getPushToken,
  getInvitationPayload,
  isDuplicateConnection,
} from '../store/store-selector'
import { saveNewConnection } from '../store/connections-store'
import { encrypt, addConnection } from '../bridge/react-native-cxs/RNCxs'
import type {
  InvitationReceivedActionData,
  InvitationResponseSendData,
  InvitationResponseSendAction,
  InvitationPayload,
  InvitationStore,
  InvitationAction,
} from './type-invitation'
import type { CustomError } from '../common/type-common'

export const invitationInitialState = {}

export const invitationReceived = (data: InvitationReceivedActionData) => ({
  type: INVITATION_RECEIVED,
  data,
})

export const sendInvitationResponse = (data: InvitationResponseSendData) => ({
  type: INVITATION_RESPONSE_SEND,
  data,
})

export const invitationSuccess = (senderDID: string) => ({
  type: INVITATION_RESPONSE_SUCCESS,
  senderDID,
})

export const invitationFail = (error: CustomError, senderDID: string) => ({
  type: INVITATION_RESPONSE_FAIL,
  error,
  senderDID,
})

export const invitationRejected = (senderDID: string) => ({
  type: INVITATION_REJECTED,
  senderDID,
})

export function* sendResponse(
  action: InvitationResponseSendAction
): Generator<*, *, *> {
  const { senderDID } = action.data
  // get data needed for agent api call from store using selectors
  // this will keep our components and screen to not pass data
  // and will keep our actions lean
  const agencyUrl: string = yield select(getAgencyUrl)
  const pushToken: string = yield select(getPushToken)
  const payload: InvitationPayload = yield select(
    getInvitationPayload,
    senderDID
  )
  const alreadyExist: boolean = yield select(isDuplicateConnection, senderDID)

  if (alreadyExist) {
    yield put(invitationFail(ERROR_ALREADY_EXIST, senderDID))
  } else {
    const metadata = {
      ...payload,
    }
    // TODO: This will remain hard coded and it should come from config store
    // we need to switch this as well while switching environment
    const agencyDid = '5qiK8KZQ86XjcnLmy5S2Tn'
    const agencyVerificationKey = '3dzsPMyBeJiGtsxWoyrfXZL6mqj3iXxdJ75vewJ1jSwn'
    const { identifier, verificationKey } = yield call(
      addConnection,
      agencyDid,
      agencyVerificationKey,
      metadata
    )

    try {
      const connectResponse = yield call(connectWithConsumerAgency, {
        agencyUrl,
        dataBody: {
          type: API_TYPE.CONNECT,
          fromDID: identifier,
          fromDIDVerKey: verificationKey,
        },
      })

      const registerResponse = yield call(registerWithConsumerAgency, {
        agencyUrl,
        dataBody: {
          type: API_TYPE.REGISTER,
          fromDID: identifier,
        },
      })

      const createAgentResponse = yield call(createAgentWithConsumerAgency, {
        agencyUrl,
        dataBody: {
          type: API_TYPE.CREATE_AGENT,
          forDID: identifier,
        },
      })

      const pairwiseConnection = yield call(
        addConnection,
        senderDID,
        payload.senderVerificationKey,
        metadata
      )

      const createPairwiseKey = yield call(createAgentPairwiseKey, {
        agencyUrl,
        dataBody: {
          to: identifier,
          agentPayload: JSON.stringify({
            type: API_TYPE.CREATE_KEY,
            forDID: pairwiseConnection.identifier,
            forDIDVerKey: pairwiseConnection.verificationKey,
            nonce: '12121212',
          }),
        },
      })

      const dataBody = {
        to: pairwiseConnection.identifier,
        agentPayload: JSON.stringify({
          type: API_TYPE.INVITE_ANSWERED,
          uid: payload.requestId,
          // TODO: get this data from agent registration
          keyDlgProof: 'delegate to agent',
          senderName: payload.senderName,
          senderLogoUrl: payload.senderLogoUrl,
          senderDID,
          senderDIDVerKey: payload.senderVerificationKey,
          remoteAgentKeyDlgProof: payload.senderAgentKeyDelegationProof,
          remoteEndpoint: payload.senderEndpoint,
          pushComMethod: `FCM:${pushToken}`,
        }),
      }

      // TODO:KS Check errors from backend in api utils
      yield call(sendInvitationResponseApi, {
        agencyUrl,
        dataBody,
      })

      yield put(invitationSuccess(senderDID))

      if (action.data.response === ResponseType.accepted) {
        const connection = {
          newConnection: {
            identifier: pairwiseConnection.identifier,
            logoUrl: payload.senderLogoUrl,
            ...payload,
          },
        }
        yield put(saveNewConnection(connection))
      }
    } catch (e) {
      let error: CustomError = {
        code: ERROR_INVITATION_RESPONSE_PARSE_CODE,
        message: ERROR_INVITATION_RESPONSE_PARSE,
      }
      try {
        error = JSON.parse(e.message)
      } catch (_) {}
      yield put(invitationFail(error, senderDID))
    }
  }
}

function* watchSendInvitationResponse() {
  yield takeLatest(INVITATION_RESPONSE_SEND, sendResponse)
}

export function* watchInvitation(): Generator<*, *, *> {
  yield all([watchSendInvitationResponse()])
}

export default function invitationReducer(
  state: InvitationStore = invitationInitialState,
  action: InvitationAction
) {
  switch (action.type) {
    case INVITATION_RECEIVED:
      return {
        ...state,
        [action.data.payload.senderDID]: {
          ...action.data,
          status: ResponseType.none,
          isFetching: false,
          error: null,
        },
      }

    case INVITATION_RESPONSE_SEND:
      return {
        ...state,
        [action.data.senderDID]: {
          ...state[action.data.senderDID],
          isFetching: true,
          status: action.data.response,
        },
      }

    case INVITATION_RESPONSE_SUCCESS:
      return {
        ...state,
        [action.senderDID]: {
          ...state[action.senderDID],
          isFetching: false,
          error: null,
        },
      }

    case INVITATION_RESPONSE_FAIL:
      return {
        ...state,
        [action.senderDID]: {
          ...state[action.senderDID],
          isFetching: false,
          error: action.error,
          status: ResponseType.none,
        },
      }

    case INVITATION_REJECTED:
      return {
        ...state,
        [action.senderDID]: {
          ...state[action.senderDID],
          isFetching: false,
          error: null,
          status: ResponseType.rejected,
        },
      }

    default:
      return state
  }
}
