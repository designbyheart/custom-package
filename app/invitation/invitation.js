import React, { Component } from "react";
import { View, StyleSheet } from "react-native";
import { Avatar, Text, Icon, Button } from "react-native-elements";
import InvitationText from "./invite-text";
import InvitationActions from "./invitation-actions";

export default class InvitationScreen extends Component {
  static navigationOptions = {
    title: "Accept invitation"
  };

  render() {
    return (
      <View style={styles.container}>
        <View style={styles.inviteContainer}>
          <InvitationText />
        </View>
        <View>
          <InvitationActions />
        </View>
      </View>
    );
  }
}

const styles = StyleSheet.create({
  container: {
    flex: 1,
    flexDirection: "column",
    justifyContent: "space-between"
  },
  inviteContainer: {
    flex: 1,
    justifyContent: "center",
    alignItems: "center"
  }
});
