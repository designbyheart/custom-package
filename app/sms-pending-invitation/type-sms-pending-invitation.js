// @flow
import type { CustomError, InitialTestAction } from '../common/type-common'

export const SMSPendingInvitationStatus = {
  NONE: 'NONE',
  FETCH_FAILED: 'FETCH_FAILED',
  RECEIVED: 'RECEIVED',
  SEEN: 'SEEN',
}

export type SMSPendingInvitationStatusType = $Keys<
  typeof SMSPendingInvitationStatus
>

export type AgentKeyDelegationProof = {
  agentDID: string,
  agentDelegatedKey: string,
  signature: string,
}

export type InvitationSenderDetail = {
  name: string,
  agentKeyDlgProof: AgentKeyDelegationProof,
  DID: string,
  logoUrl: string,
  verKey: string,
}

export type InvitationSenderAgencyDetail = {
  DID: string,
  verKey: string,
  endpoint: string,
}

export type SMSPendingInvitationPayload = {
  connReqId: string,
  statusCode: string,
  senderDetail: InvitationSenderDetail,
  senderAgencyDetail: InvitationSenderAgencyDetail,
  targetName: string,
}

export type SMSPendingInvitationStore = {
  +payload: ?SMSPendingInvitationPayload,
  +status: SMSPendingInvitationStatusType,
  +isFetching: boolean,
  +error?: ?CustomError,
}

export const SMS_PENDING_INVITATION_REQUEST: 'SMS_PENDING_INVITATION_REQUEST' =
  'SMS_PENDING_INVITATION_REQUEST'

export type SMSPendingInvitationRequestAction = {
  type: typeof SMS_PENDING_INVITATION_REQUEST,
  smsToken: string,
}

export const SMS_PENDING_INVITATION_RECEIVED: 'SMS_PENDING_INVITATION_RECEIVED' =
  'SMS_PENDING_INVITATION_RECEIVED'

export type SMSPendingInvitationReceivedAction = {
  type: typeof SMS_PENDING_INVITATION_RECEIVED,
  data: SMSPendingInvitationPayload,
}

export const SMS_PENDING_INVITATION_FAIL: 'SMS_PENDING_INVITATION_FAIL' =
  'SMS_PENDING_INVITATION_FAIL'

export type SMSPendingInvitationFailAction = {
  type: typeof SMS_PENDING_INVITATION_FAIL,
  error: CustomError,
}

export const SMS_PENDING_INVITATION_SEEN: 'SMS_PENDING_INVITATION_SEEN' =
  'SMS_PENDING_INVITATION_SEEN'

export type SMSPendingInvitationSeenAction = {
  type: typeof SMS_PENDING_INVITATION_SEEN,
}

export type SMSPendingInvitationAction =
  | SMSPendingInvitationRequestAction
  | SMSPendingInvitationReceivedAction
  | SMSPendingInvitationFailAction
  | SMSPendingInvitationSeenAction
  | InitialTestAction
