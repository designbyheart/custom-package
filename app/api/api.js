// @flow
import { api, options } from './api-utils'
import type {
  GetInvitationLinkApiData,
  MessageApiData,
  DownloadInvitationApiData,
  SendAuthenticationResponseApiData,
  GetProfileApiData,
  SendMessageApiData,
  SendClaimRequestApiData,
  ApiClaimRequest,
  EdgeClaimRequest,
} from './type-api'
import { PAYLOAD_TYPE, MESSAGE_TYPE, STATUS_CODE } from './api-constants'

export const getInvitationLink = ({
  agencyUrl,
  smsToken,
}: GetInvitationLinkApiData) =>
  api(`${agencyUrl}/agency/url-mapper/${smsToken}`, options('GET'))

export const invitationDetailsRequest = ({ url }: DownloadInvitationApiData) =>
  api(url, options('GET'))

export const connectWithConsumerAgency = ({
  agencyUrl,
  dataBody,
}: MessageApiData) => api(`${agencyUrl}/agency/msg`, options('POST', dataBody))

export const registerWithConsumerAgency = ({
  agencyUrl,
  dataBody,
}: MessageApiData) => api(`${agencyUrl}/agency/msg`, options('POST', dataBody))

export const createAgentWithConsumerAgency = ({
  agencyUrl,
  dataBody,
}: MessageApiData) => api(`${agencyUrl}/agency/msg`, options('POST', dataBody))

export const createAgentPairwiseKey = ({
  agencyUrl,
  dataBody,
}: MessageApiData) =>
  api(`${agencyUrl}/agency/route`, options('POST', dataBody))

export const sendInvitationResponse = ({
  agencyUrl,
  dataBody,
}: MessageApiData) =>
  api(`${agencyUrl}/agency/route`, options('POST', dataBody))

export const getAdditionalData = ({ agencyUrl, dataBody }: MessageApiData) =>
  api(`${agencyUrl}/agency/route`, options('POST', dataBody))

export const sendAuthenticationRequest = ({
  data: { identifier, dataBody },
  config: { agencyUrl },
}: SendAuthenticationResponseApiData) =>
  api(`${agencyUrl}/agent/${identifier}/auth`, options('PUT', dataBody), true)

export const getProfile = ({
  identifier,
  challenge,
  signature,
  agencyUrl,
}: GetProfileApiData) =>
  api(
    `${agencyUrl}/agent/${identifier}/remote/profile?challenge=${challenge}&signature=${signature}`,
    options('GET')
  )

export const sendUpdatedPushToken = ({ agencyUrl, dataBody }: MessageApiData) =>
  api(`${agencyUrl}/agency/route`, options('POST', dataBody))

export const sendMessage = ({
  agencyUrl,
  userPairwiseDid,
  agentPayload,
}: SendMessageApiData) =>
  api(
    `${agencyUrl}/agency/route`,
    options('POST', {
      to: userPairwiseDid,
      agentPayload: JSON.stringify(agentPayload),
    })
  )

export function convertClaimRequestToEdgeClaimRequest(
  claimRequest: ApiClaimRequest
): EdgeClaimRequest {
  const { blinded_ms, schema_seq_no, issuer_did } = claimRequest

  return {
    blinded_ms,
    issuer_did,
    schema_seq_no,
    msg_type: MESSAGE_TYPE.CLAIM_REQUEST,
    // hard coded version as of now, update once versioning is implemented
    version: '0.1',
    to_did: claimRequest.remoteDid,
    from_did: claimRequest.userPairwiseDid,
    tid: '1',
    mid: '1',
  }
}

export const sendClaimRequest = ({
  agencyUrl,
  userPairwiseDid,
  responseMsgId,
  claimRequest,
}: SendClaimRequestApiData) =>
  sendMessage({
    agencyUrl,
    userPairwiseDid,
    agentPayload: {
      type: PAYLOAD_TYPE.SEND_MESSAGE,
      msgType: MESSAGE_TYPE.CLAIM_REQUEST,
      edgeAgentPayload: JSON.stringify(
        convertClaimRequestToEdgeClaimRequest(claimRequest)
      ),
      refMsgId: responseMsgId,
      // just accept for now, we are not sending rejection as of now
      statusCode: STATUS_CODE.ACCEPT,
    },
  })
