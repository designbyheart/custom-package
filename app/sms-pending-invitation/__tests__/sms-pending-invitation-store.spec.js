// @flow
import { call, select, put } from 'redux-saga/effects'
import { getAgencyUrl } from '../../store/store-selector'
import smsPendingInvitationReducer, {
  callSmsPendingInvitationRequest,
  convertSmsPayloadToInvitation,
  smsPendingInvitationFail,
  smsPendingInvitationReceived,
  getSmsPendingInvitation,
  smsPendingInvitationSeen,
} from '../sms-pending-invitation-store'
import { SMSPendingInvitationStatus } from '../type-sms-pending-invitation'
import { getInvitationLink, invitationDetailsRequest } from '../../api/api'
import { PENDING_CONNECTION_REQUEST_CODE } from '../../api/api-constants'
import { initialTestAction } from '../../common/type-common'
import { invitationReceived } from '../../invitation/invitation-store'

describe('SMS Connection Request store', () => {
  const initialState = {
    payload: null,
    status: SMSPendingInvitationStatus.NONE,
    isFetching: false,
    error: null,
  }
  const smsToken = 'gm76ku'

  const payload = {
    statusCode: 'MS-102',
    connReqId: 'nzc5ztz',
    senderDetail: {
      name: 'default',
      agentKeyDlgProof: {
        agentDID: 'EHNMj5FNjA3xBo8HUKQTZv',
        agentDelegatedKey: '8Esrno7vguYbwVJDa31aMCdWJTu2tC7tke8u2iKUUvgB',
        signature:
          'OXtzqX9LJJsBY/Mrhjff6b9L79Pdg8U4B7dI/3RR65BJr5jktAGMwpMLe1aQRQT0FlplZfNy8QtUd4KNPd+ZAg==',
      },
      DID: '3Bb6zCYMoGnMTho97HGDrn',
      logoUrl: 'http://www.evernym.com',
      verKey: '2BzhMgbLXbECPPYiLtSEd36Boxt1oquuGJeS1ofhPoEr',
    },
    senderAgencyDetail: {
      DID: 'BDSmVkzxRYGE4HKyMKxd1H',
      verKey: '6yUatReYWNSUfEtC2ABgRXmmLaxCyQqsjLwv2BomxsxD',
      endpoint: '52.38.32.107:80/agency/msg',
    },
    targetName: 'there',
    statusMsg: 'message sent',
  }

  const getPendingInvitationState = state =>
    smsPendingInvitationReducer(state, getSmsPendingInvitation(smsToken))

  it('should be correct initial state', () => {
    const actualInitialState = smsPendingInvitationReducer(
      undefined,
      initialTestAction()
    )
    expect(actualInitialState).toEqual(initialState)
  })

  it('should update store when pending invitation is requested', () => {
    const nextState = getPendingInvitationState(initialState)

    expect(nextState).toMatchSnapshot()
  })

  it('should update store when invitation received', () => {
    const invitationRequestedState = getPendingInvitationState(initialState)
    const nextState = smsPendingInvitationReducer(
      invitationRequestedState,
      smsPendingInvitationReceived(payload)
    )

    expect(nextState).toMatchSnapshot()
  })

  it('should update store when invitation receive fail', () => {
    const invitationRequestedState = getPendingInvitationState(initialState)
    const nextState = smsPendingInvitationReducer(
      invitationRequestedState,
      smsPendingInvitationFail({
        code: 'TEST-FAIL',
        message: 'Test fail message',
      })
    )

    expect(nextState).toMatchSnapshot()
  })

  it('should update store when invitation is seen', () => {
    const invitationRequestedState = getPendingInvitationState(initialState)
    const afterReceived = smsPendingInvitationReducer(
      invitationRequestedState,
      smsPendingInvitationReceived(payload)
    )
    const nextState = smsPendingInvitationReducer(
      afterReceived,
      smsPendingInvitationSeen()
    )

    expect(nextState).toMatchSnapshot()
  })

  it('sms invitation download workflow should work fine if api returns success', () => {
    const gen = callSmsPendingInvitationRequest(
      getSmsPendingInvitation(smsToken)
    )

    expect(gen.next().value).toEqual(select(getAgencyUrl))
    const agencyUrl = 'http://test-agency.com'

    expect(gen.next(agencyUrl).value).toEqual(
      call(getInvitationLink, {
        agencyUrl,
        smsToken,
      })
    )

    const invitationData = { url: 'http://enterprise-agency.com' }

    expect(gen.next(invitationData).value).toEqual(
      call(invitationDetailsRequest, {
        url: invitationData.url,
      })
    )

    expect(gen.next(payload).value).toEqual(
      put(smsPendingInvitationReceived(payload))
    )

    expect(gen.next().value).toEqual(
      put(
        invitationReceived({ payload: convertSmsPayloadToInvitation(payload) })
      )
    )

    expect(gen.next().done).toBe(true)
  })

  it('sms invitation download error should raise fail action', () => {
    const error = {
      code: 'OCS',
      message: 'sms connection pending request api error',
    }

    const gen = callSmsPendingInvitationRequest(
      getSmsPendingInvitation(smsToken)
    )

    expect(gen.next().value).toEqual(select(getAgencyUrl))
    const agencyUrl = 'http://test-agency.com'

    expect(gen.next(agencyUrl).value).toEqual(
      call(getInvitationLink, {
        agencyUrl,
        smsToken,
      })
    )

    expect(gen.throw(new Error(JSON.stringify(error))).value).toEqual(
      put(smsPendingInvitationFail(error))
    )

    expect(gen.next().done).toBe(true)
  })
})
