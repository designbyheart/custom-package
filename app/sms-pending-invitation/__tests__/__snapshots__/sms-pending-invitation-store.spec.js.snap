// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`SMS Connection Request store should update store when invitation is seen 1`] = `
Object {
  "error": null,
  "isFetching": false,
  "payload": Object {
    "connReqId": "nzc5ztz",
    "senderAgencyDetail": Object {
      "DID": "BDSmVkzxRYGE4HKyMKxd1H",
      "endpoint": "52.38.32.107:80/agency/msg",
      "verKey": "6yUatReYWNSUfEtC2ABgRXmmLaxCyQqsjLwv2BomxsxD",
    },
    "senderDetail": Object {
      "DID": "3Bb6zCYMoGnMTho97HGDrn",
      "agentKeyDlgProof": Object {
        "agentDID": "EHNMj5FNjA3xBo8HUKQTZv",
        "agentDelegatedKey": "8Esrno7vguYbwVJDa31aMCdWJTu2tC7tke8u2iKUUvgB",
        "signature": "OXtzqX9LJJsBY/Mrhjff6b9L79Pdg8U4B7dI/3RR65BJr5jktAGMwpMLe1aQRQT0FlplZfNy8QtUd4KNPd+ZAg==",
      },
      "logoUrl": "http://www.evernym.com",
      "name": "default",
      "verKey": "2BzhMgbLXbECPPYiLtSEd36Boxt1oquuGJeS1ofhPoEr",
    },
    "statusCode": "MS-102",
    "statusMsg": "message sent",
    "targetName": "there",
  },
  "status": "SEEN",
}
`;

exports[`SMS Connection Request store should update store when invitation receive fail 1`] = `
Object {
  "error": Object {
    "code": "TEST-FAIL",
    "message": "Test fail message",
  },
  "isFetching": false,
  "payload": null,
  "status": "FETCH_FAILED",
}
`;

exports[`SMS Connection Request store should update store when invitation received 1`] = `
Object {
  "error": null,
  "isFetching": false,
  "payload": Object {
    "connReqId": "nzc5ztz",
    "senderAgencyDetail": Object {
      "DID": "BDSmVkzxRYGE4HKyMKxd1H",
      "endpoint": "52.38.32.107:80/agency/msg",
      "verKey": "6yUatReYWNSUfEtC2ABgRXmmLaxCyQqsjLwv2BomxsxD",
    },
    "senderDetail": Object {
      "DID": "3Bb6zCYMoGnMTho97HGDrn",
      "agentKeyDlgProof": Object {
        "agentDID": "EHNMj5FNjA3xBo8HUKQTZv",
        "agentDelegatedKey": "8Esrno7vguYbwVJDa31aMCdWJTu2tC7tke8u2iKUUvgB",
        "signature": "OXtzqX9LJJsBY/Mrhjff6b9L79Pdg8U4B7dI/3RR65BJr5jktAGMwpMLe1aQRQT0FlplZfNy8QtUd4KNPd+ZAg==",
      },
      "logoUrl": "http://www.evernym.com",
      "name": "default",
      "verKey": "2BzhMgbLXbECPPYiLtSEd36Boxt1oquuGJeS1ofhPoEr",
    },
    "statusCode": "MS-102",
    "statusMsg": "message sent",
    "targetName": "there",
  },
  "status": "RECEIVED",
}
`;

exports[`SMS Connection Request store should update store when pending invitation is requested 1`] = `
Object {
  "error": null,
  "isFetching": true,
  "payload": null,
  "status": "NONE",
}
`;
